### PR TITLE
Skill Die and Grid Power Progression Added

### PR DESCRIPTION
**Pull request by JFranklin3000**
##### In this PR
- Added Skill Die option to Roles and Grid Power Progression

##### Testing
- `fvtt package pack -n "prcrbitems"`
- Validated that the Skill Die is being set correctly 
- Validate that the skill Die can be selected on weapon effects. 
- figure out why the isSpecialization isn't working.

### CODE REVIEW

File: packs/prcrbitems/_source/White_Ranger_6ANY0ZSPWC9G1WHD.json [hunk 1]
```diff
@@ -133,36 +167,44 @@
...
-        "level20"
-      ],
-      "speed": [
-        "level4",
-        "level8",
-        "level12"
+        "level19"
       ],
-      "smarts": [
+      "specialization": [
```
Comment (phildominguez) [inline packs/prcrbitems/_source/White_Ranger_6ANY0ZSPWC9G1WHD.json:187]: This is currently `specialized` in the template

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/helpers/utils.mjs [hunk 2]
```diff
@@ -521,6 +520,40 @@ export async function setRoleValues(role, actor, newLevel=null, previousLevel=nu
     });
   }
 
+  if (role.system.skillDie.has) {
+    const skillName = role.system.skillDie.name.toLowerCase().replace(/[^a-zA-Z0-9]+(.)/g, (m, chr) => chr.toUpperCase());
+    const shiftList=CONFIG.E20.skillShiftList;
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/helpers/utils.mjs:525]: fixed

File: templates/actor/parts/items/weapon/container.hbs [hunk 3]
```diff
@@ -28,8 +28,12 @@
           </div>
           <select class="inline-edit" data-field="system.items.{{@key}}.classification.skill" name="item.classification.skill">
             {{#select item.classification.skill}}
-            {{#each @root.config.skills as |name type|}}
-            <option value="{{type}}">{{name}}</option>
+            {{#each @root.actor.system.skills}}
+              {{#if this.displayName}}
+                <option value="{{@key}}">{{this.displayName}}</option>
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/actor/parts/items/weapon/container.hbs:33]: It is in this case, there was a lot of research into this one. :) 
https://handlebarsjs.com/guide/builtin-helpers.html#each

File: template.json [hunk 4]
```diff
@@ -890,6 +895,12 @@
           "starting": 0
         }
       },
+      "skillDie": {
+        "has": false,
+        "name": "",
+        "levels": [],
+        "specialized": []
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply template.json:902]: Changed that makes sense.

File: template.json [hunk 5]
```diff
@@ -890,6 +895,12 @@
           "starting": 0
         }
       },
+      "skillDie": {
+        "has": false,
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply template.json:899]: fixed

File: templates/item/sheets/role.hbs [hunk 6]
```diff
@@ -233,24 +233,95 @@
         {{/"systems/essence20/templates/item/parts/sheet-field.hbs"}}
     {{/ifNotEquals}}
 
-      {{!-- General Perk --}}
-      {{#> "systems/essence20/templates/item/parts/sheet-field.hbs" label='E20.PerkGeneralPlural'}}
+    {{!-- Role has SkillDie --}}
+    {{#> "systems/essence20/templates/item/parts/sheet-field.hbs" label='E20.hasSkillDie'}}
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/item/sheets/role.hbs:237]: Changed to match other change

File: packs/prcrbitems/_source/White_Ranger_6ANY0ZSPWC9G1WHD.json (hunk 1 above)
Comment (JFranklin3000) [reply packs/prcrbitems/_source/White_Ranger_6ANY0ZSPWC9G1WHD.json:187]: Changed to specializedLevels

File: module/helpers/utils.mjs [hunk 7]
```diff
@@ -521,6 +520,40 @@ export async function setRoleValues(role, actor, newLevel=null, previousLevel=nu
...
+    const finalShiftIndex = Math.max(
+      0,
+      Math.min(shiftList.length - 1, initialShiftIndex-totalChange),
+    );
+
+    const skillStringShift = `system.skills.${skillName}.shift`;
+    const skillStringIsSpecialized = `system.skills.${skillName}.isSpecialized`;
+
+    let isSpecialized = false;
+    if (role.system.skillDie.specialization) {
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/helpers/utils.mjs:541]: Why would we remove the if? it is preventing an error I think as well.

File: templates/actor/parts/items/weapon/container.hbs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/actor/parts/items/weapon/container.hbs:33]: see the object section.

File: module/sheet-handlers/role-handler.mjs [hunk 8]
```diff
@@ -152,6 +152,38 @@ export class RoleHandler {
...
+      const skillStringEssencesSocial = `system.skills.${skillName}.essences.social`;
+      const skillStringEssencesSpeed = `system.skills.${skillName}.essences.speed`;
+      const skillStringEssencesSrength = `system.skills.${skillName}.essences.strength`;
+      const skillStringIsSpecialized = `system.skills.${skillName}.isSpecialized`;
+      const skillStringModifier = `system.skills.${skillName}.modifier`;
+      const skillStringSnag = `system.skills.${skillName}.snag`;
+      const skillStringShiftUp = `system.skills.${skillName}.shiftUp`;
+      const skillStringShiftDown = `system.skills.${skillName}.shiftDown`;
+
+
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs]: fixed

File: module/helpers/roll-dialog.mjs [hunk 9]
```diff
@@ -31,6 +31,13 @@ export class RollDialog {
    * @returns {Promise<Dialog>}   The dialog to be displayed.
    */
   async getSkillRollOptions(dataset, skillDataset, actor) {
+    let specializationBoolean = false;
+    if (dataset.isSpecialized === true || dataset.isSpecialized === false) {
+      specializationBoolean = dataset.isSpecialized;
+    } else if (dataset.isSpecialized === 'true') {
+      specializationBoolean = true;
+    }
+
```
Comment (phildominguez) [inline module/helpers/roll-dialog.mjs:40]:
```suggestion
```

File: module/dice.mjs [hunk 10]
```diff
@@ -65,19 +65,36 @@ export class Dice {
    * @param {Item} item   The item being used, if any.
    */
   async rollSkill(rawDataset, actor, item) {
+    let specializationBoolean = false;
+    if (rawDataset.isSpecialized === true || rawDataset.isSpecialized === false) {
+      specializationBoolean = rawDataset.isSpecialized;
+    } else if (rawDataset.isSpecialized === 'true') {
+      specializationBoolean = true;
+    }
+
```
Comment (phildominguez) [inline module/dice.mjs:74]:
```suggestion
```

File: module/dice.mjs [hunk 11]
```diff
@@ -65,19 +65,36 @@ export class Dice {
...
     const dataset = { // Converting strings to usable types
       ...rawDataset,
-      isSpecialized: rawDataset.isSpecialized === 'true',
+      isSpecialized: specializationBoolean,
```
Comment (phildominguez) [inline module/dice.mjs:77]:
```suggestion
      isSpecialized: dataset.isSpecialized,
```

File: module/helpers/utils.mjs [hunk 12]
```diff
@@ -521,6 +520,39 @@ export async function setRoleValues(role, actor, newLevel=null, previousLevel=nu
...
+    let isSpecialized = false;
+    for (const arrayLevel of role.system.skillDie.specializedLevels) {
+      const level = arrayLevel.replace(/[^0-9]/g, '');
+      if (actor.system.level == level) {
+        isSpecialized = true;
+        break;
+      }
+    }
+
+
```
Comment (phildominguez) [inline module/helpers/utils.mjs:549]:
```suggestion
```

File: templates/item/sheets/role.hbs [hunk 13]
```diff
@@ -233,24 +233,95 @@
...
+      {{#*inline "item-field-inputs"}}
+      <div class="flex-group-right chip-section">
+      {{#each item.system.skillDie.levels as |name|}}
+        <span class="chip" name="chip.skillDie.levels.{{type}}">{{lookup @root.config.actorLevels name}}</span>
+      {{/each}}
+      </div>
+      {{/inline}}
+      {{/"systems/essence20/templates/item/parts/sheet-field.hbs"}}
+
+      {{#> "systems/essence20/templates/item/parts/sheet-field.hbs" label='E20.SkillDieSpecializedLevels'}}
```
Comment (phildominguez) [inline templates/item/sheets/role.hbs:268]: `E20.SkillDieSpecializedLevels` not defined

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: templates/item/sheets/role.hbs (hunk 13 above)
Comment (JFranklin3000) [reply templates/item/sheets/role.hbs:268]: Or in this case it was actually a typo of SillDieSpecializedLevels

File: module/helpers/roll-dialog.mjs [hunk 14]
```diff
@@ -41,7 +41,7 @@ export class RollDialog {
       {
         shiftUp: dataset.shiftUp || 0,
         shiftDown: dataset.shiftDown || 0,
-        isSpecialized: dataset.isSpecialized === 'true',
+        isSpecialized: dataset.isSpecialized,
```
Comment (phildominguez) [inline module/helpers/roll-dialog.mjs:44]:
```suggestion
```

File: module/dice.mjs [hunk 15]
```diff
@@ -67,17 +67,27 @@ export class Dice {
   async rollSkill(rawDataset, actor, item) {
     const dataset = { // Converting strings to usable types
       ...rawDataset,
-      isSpecialized: rawDataset.isSpecialized === 'true',
+      isSpecialized: dataset.isSpecialized,
```
Comment (phildominguez) [inline module/dice.mjs:70]:
```suggestion
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: lang/en.json [hunk 16]
```diff
@@ -380,6 +380,7 @@
     "RoleDrop": "Drop a Role here",
     "RoleMultipleError": "Characters can only have one Role",
     "RolePerks": "Role Perks",
+    "RoleSkillDieError": "The roles Skill Die has no name.",
```
Comment (phildominguez) [inline lang/en.json:383]:
```suggestion
    "RoleSkillDieError": "Dropped Role's Skill Die must have a name.",
```

File: module/sheet-handlers/role-handler.mjs [hunk 17]
```diff
@@ -150,6 +150,11 @@ export class RoleHandler {
       return false;
     }
 
+    if (role.system.skillDie.isUsed && role.system.skillDie.name == "") {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:153]:
```suggestion
    if (role.system.skillDie.isUsed && !role.system.skillDie.name) {
```

File: module/sheet-handlers/role-handler.mjs [hunk 18]
```diff
@@ -150,6 +150,11 @@ export class RoleHandler {
...
+      ui.notifications.error(game.i18n.format(game.i18n.localize('E20.RoleSkillDieError')));
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:154]:
```suggestion
      ui.notifications.error(game.i18n.localize('E20.RoleSkillDieError'));
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

**Review by phildominguez — APPROVED** [review]

**Pull request merged**